src/getkey.1
.\" A man page for getkey(1). -*- nroff -*-
.\"
.\" Copyright (C) 2006 Red Hat, Inc. All rights reserved.
.\"
.\" This copyrighted material is made available to anyone wishing to use,
.\" modify, copy, or redistribute it subject to the terms and conditions of the
.\" GNU General Public License v.2.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
.\" FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for 
.\" more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
.\"
.\" Author: Miloslav Trmac <mitr@redhat.com>
.TH getkey 1 "Jan 2006"

.SH NAME
getkey \- wait until a key is pressed

.SH SYNOPSIS
\fBgetkey\fR [\fIOPTION\fR]... [\fIKEYS\fR]

.SH DESCRIPTION
.B getkey
waits until one of
.I KEYS
is pressed.
If
.I KEYS
are not specified, any key is accepted.
.I KEYS
are matched case-insensitive.

.SH EXIT STATUS
.B getkey
exits with status 0 if one of the expected keys is pressed.
If invalid arguments are specified,
.B getkey
exits with status 255.
If
.B getkey
is interrupted or the wait times out,
.B getkey
exits with other non-zero status.

.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-wait\fR \fISECONDS\fR
Wait only for
.I SECONDS
seconds.
The default is 0, which means to wait without a time limit.

.TP
\fB\-i\fR, \fB\-\-ignore\-control\-chars\fR
Don't treat Ctrl+C and Ctrl+D specially.
When this option is not specified, these characters interrupt \fBgetkey\fR.

.TP
\fB\-m\fR, \fB\-\-message\fR \fIMESSAGE\fR
Display
.I MESSAGE
while waiting.
The message is used as a format string in
.BR sprintf (3),
with a single argument, the number of seconds left.
Typical usage is therefore
\fB"Press a key within %d seconds to ..."\fR.
If
.I MESSAGE
contains other format string directives, the behavior is undefined and
.B getkey
may crash.

If there is no time limit specified,
the number of seconds left is reported as 0.


ChangeLog
2006-06-12  Bill Nottingham  <notting@redhat.com>

	* ChangeLog, initscripts.spec: 8.35-1

	* rc.d/rc.sysinit, sysconfig/readonly-root:
	readonly root enhancments (modified from <law@redhat.com>, #193164)

2006-06-12  Subhransu Behera  <sbehera@redhat.com>

	* po/or.po: Updated Oriya translation

2006-06-11  Sharuzzaman Ahmat Raslan <sharuzzaman@myrealbox.com>

	* po/ms.po: Updated Malay translation

2006-06-10  Subhransu Behera  <sbehera@redhat.com>

	* po/or.po: Updated Oriya translation

2006-06-09  Dimitris Michelinakis  <dimitris@michelinakis.gr>

	* po/el.po: Minor changes to Greek translation

2006-06-08  Subhransu Behera  <sbehera@redhat.com>

	* po/or.po: Updated Oriya translation

2006-06-08  Dimitris Michelinakis  <dimitris@michelinakis.gr>

	* po/el.po: Minor change in Greek translation

2006-06-07  Subhransu Behera  <sbehera@redhat.com>

	* po/or.po: Updated Oriya translation

2006-06-07  Miloslav Trmac  <mitr@redhat.com>

	* Makefile, crypttab.5: Add /etc/crypttab documentation.

2006-06-06  Dimitris Michelinakis  <dimitris@michelinakis.gr>

	* po/el.po: More Greek translation updates to initscripts

2006-06-05  Miloslav Trmac  <mitr@redhat.com>

	* rc.d/init.d/functions: Clarify killproc usage message (#193711)

2006-06-03  Piotr Drąg  <raven@pmail.pl>

	* po/pl.po: Updated and improved Polish translation

2006-06-02  Dimitris Michelinakis  <dimitris@michelinakis.gr>

	* po/el.po: More updates to initscripts Greek translation

2006-06-01  Dimitris Michelinakis  <dimitris@michelinakis.gr>

	* po/el.po: First steps to translate initscripts to Greek

2006-06-01  Miloslav Trmac  <mitr@redhat.com>

	* rc.d/init.d/halt, rc.d/rc.sysinit:
	Add encrypted swap and non-root filesystem support (#127378, based on
	a patch by W. Michael Petullo <redhat@flyn.org> and Debian's cryptsetup
	package)

2006-05-30  Miloslav Trmac  <mitr@redhat.com>

	* sysconfig.txt: Document BRIDGE= (#192576)

2006-05-29  Joel Gump  <biglaughing@gmail.com>

	* po/zh_CN.po: re_translate fuzzy words...of the current po file

2006-05-25  Bill Nottingham  <notting@redhat.com>

	* prefdm: use the right option (#193177, <david@lovesunix.net>)

	* rc.d/rc.sysinit: googlespell prefers one l. words takes both.

	* rc.d/rc.sysinit:
	allow for sulogin instead of automatic relabeling (<dwalsh@redhat.com>)

2006-05-23  Bill Nottingham  <notting@redhat.com>

	* initscripts.spec, ChangeLog: 8.34-1

	* initscripts-s390.patch, initscripts.spec, rc.d/init.d/halt, rc.d/rc.sysinit:
	integrate s390 patch - just check for hwclock before running it

	* rc.d/rc.sysinit: handle cups (#189168)

	* initscripts.spec, src/Makefile:
	link glib2 dynamically again now that it's in /lib

2006-05-22  Sharuzzaman Ahmat Raslan <sharuzzaman@myrealbox.com>

	* po/ms.po: Updated Malay translation

2006-05-22  Amanpreet Brar  <aalam@redhat.com>

	* po/kn.po, po/te.po: adding new files for language Telugu/Kannada

2006-05-17  Miloslav Trmac  <mitr@redhat.com>

	* initscripts.spec, sysconfig/network-scripts/ifdown-aliases, sysconfig/network-scripts/ifup-aliases:
	Remove unused ifdown-aliases.

	* src/ppp-watch.c: Fix typo.

2006-05-17  Bill Nottingham  <notting@redhat.com>

	* rwtab: add some more bits from jorton

2006-05-16  Bill Nottingham  <notting@redhat.com>

	* initscripts.spec: BuildPreReq -> BuildRequires

2006-05-16  Miloslav Trmac  <mitr@redhat.com>

	* sysconfig/network-scripts/ifdown-ipsec, sysconfig/network-scripts/ifup-ipsec:
	* ifup-ipsec:
	  - Fix key handling when AH or ESP is not used (#166257, patch by
	    Tarhon-Onu Victor <mituc@iasi.rdsnet.ro>)
	  - Allow manual tunnel mode without using AH or ESP
	* ifdown-ipsec:
	  - Fix syntax errors in manual mode when AH or ESP is not used

	* sysconfig/network-scripts/network-functions-ipv6:
	Fix IPv6 address validity check not to consider every non-empty string valid.

2006-05-14  Miloslav Trmac  <mitr@redhat.com>

	* ppp/ip-down.ipv6to4, ppp/ip-up.ipv6to4, sysconfig/network-scripts/ifdown-ipv6, sysconfig/network-scripts/ifdown-sit, sysconfig/network-scripts/ifup-ipv6, sysconfig/network-scripts/ifup-sit, sysconfig/network-scripts/network-functions-ipv6, sysconfig.txt:
	Modify IPV6TO4_ROUTING to also add addresses on the interfaces
	Add RFC 3041 (IPv6 privacy) support
	(patch for both by Peter Bieringer <pb@bieringer.de>)

2006-05-10  Miloslav Trmac  <mitr@redhat.com>

	* sysconfig.txt, sysconfig/network-scripts/ifdown-post, sysconfig/network-scripts/ifup-aliases, sysconfig/network-scripts/ifup-ctc, sysconfig/network-scripts/ifup-eth, sysconfig/network-scripts/ifup-ippp, sysconfig/network-scripts/ifup-iucv, sysconfig/network-scripts/ifup-plip, sysconfig/network-scripts/ifup-plusb, sysconfig/network-scripts/ifup-ppp, sysconfig/network-scripts/network-functions:
	Add METRIC= support for default routes (#124045)

	* ifup-ppp, network-functions: Save all default routes, lose no data.
	* network-functions:
	  - (find_gateway_dev): Fix - it should find the device while $GATEWAY is not
	    the default route
	  - (add_default_route): Don't use configuration for the device that is being
	    brought down; document function purpose and bugs.

2006-05-08  Miloslav Trmac  <mitr@redhat.com>

	* initscripts.spec, rc.d/init.d/functions, rc.d/init.d/halt, rc.d/init.d/netfs, rc.d/rc.sysinit, src/Makefile, src/fstab-decode.8, src/fstab-decode.c:
	Fix handling of mount points with white space (#186713):
	- src/fstab-decode.c, src/fstab-decode.8: New files.
	- src/Makefile: Build and install fstab-decode.
	- initscripts.spec: Add fstab-decode to %files
	- rc.d/init.d/functions (fstab_decode_str): New function.
	  (__umount_loop, __umount_loopback_loop): Factor out from copy&pasted copies.
	  (action): Fix command quoting.
	- rc.d/rc.sysinit: Correctly quote and decode mount points.
	- rc.d/init.d/halt: Use __umount_* and fstab-decode.
	  (action): Rename from runcmd.  Fix command quoting.
	- rc.d/init.d/netfs: Use __umount_*.

2006-05-08  Renato Pavicic  <repavici@globalnet.hr>


src/getkey.c
/*
 * Copyright (c) 1999-2003, 2006 Red Hat, Inc. All rights reserved.
 *
 * This software may be freely redistributed under the terms of the GNU
 * public license.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * getkey
 *
 * A very simple keygrabber.
 *
 */
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>
#include <sys/poll.h>
#include "popt.h"

static struct termios orig_tp;

static void reset_term(int x) {
	tcsetattr(0,TCSANOW,&orig_tp);
	_exit(x);
}

int main(int argc, char **argv) {
        static const char default_list[] = "";

	const char *list;
	char *waitmessage = NULL;
	char *waitprint, *waitsprint;
	int waitseconds=0;
	int alarmlen=0;
	int ignore_control=0;
	struct termios tp;
	int r;
	struct pollfd ufds; /* only one, no need for an array... */
        poptContext context;
	struct poptOption options[] = {
            { "wait", 'c', POPT_ARG_INT, &waitseconds, 0, "Number of seconds to wait for keypress", NULL },
            { "message", 'm', POPT_ARG_STRING, &waitmessage, 0, "Message to print out while waiting for string\nNOTE: The message may have a \"%d\" in it, to hold the number of seconds left to wait.", NULL },
            { "ignore-control-chars", 'i', POPT_ARG_NONE, &ignore_control, 0, "Ignore Control-C and Control-D", NULL },
            POPT_AUTOHELP
            POPT_TABLEEND
        };

	context = poptGetContext("getkey", argc, (const char **)argv, options,
                                POPT_CONTEXT_POSIXMEHARDER);
        poptSetOtherOptionHelp(context, "[keys]");

        r = poptGetNextOpt(context);
        if (r < -1) {
            fprintf(stderr, "%s: %s\n", 
                   poptBadOption(context, POPT_BADOPTION_NOALIAS),
                   poptStrerror(r));

            return -1;
        }
        list = poptGetArg(context);
	if (list != NULL) {
	    char *p;

	    p = strdup(list);
	    list = p;
	    while (*p != 0) {
		*p = toupper(*p);
		p++;
	    }
	} else
	    list = default_list;
	if (waitseconds) {
	    if (waitseconds < 0) {
		fprintf(stderr, "--wait: Invalid time %d seconds\n",
			waitseconds);
		return -1;
	    }
	    alarmlen = waitseconds;
	}

	tcgetattr(0,&tp);
	orig_tp = tp;
	signal(SIGTERM,reset_term);
	if (alarmlen != 0) {
	    signal(SIGALRM,reset_term);
	    alarm(alarmlen);
	}

	tp.c_iflag=0;
	tp.c_oflag &= ~OPOST;
	tp.c_lflag &= ~(ISIG | ICANON);
	tcsetattr(0,TCSANOW,&tp);

	ufds.events = POLLIN;
	ufds.fd = 0;

	if (waitmessage) {
	    waitprint = alloca (strlen(waitmessage)+15); /* long enough */
	    waitprint[0] = '\r';
	    waitsprint = waitprint + 1;
	}

	while (1) {
	    if (waitmessage) {
		sprintf (waitsprint, waitmessage, waitseconds);
		write (1, waitprint, strlen(waitprint));
	    }
	    r = poll(&ufds, 1, alarmlen ? 1000 : -1);
	    if (r == 0) {
		/* we have waited a whole second with no keystroke... */
		waitseconds--;
	    }
	    if (r > 0) {
		char ch;

		read(0, &ch, sizeof(ch));
		ch = toupper(ch);
		/* Die if we get a control-c or control-d */
                if (ignore_control == 0 && (ch == 3 || ch == 4))
		    reset_term(1);
		/* Don't let a null character be interpreted as a match
		   by strchr */
		if (ch != 0
		    && (strcmp(list, "") == 0 || strchr(list, ch) != NULL))
		    reset_term(0);
	    }
	}
}
